File: packages/sdk-swift/Sources/Rovenue/PaywallUI/PaywallOverrides.swift
//  PaywallOverrides.swift — Swift port of
//  packages/shared/src/paywall/validate.ts's `applyOverrides`: same
//  semantics (array order, later wins, shallow per-field overlay, identity
//  when nothing matches, unknown `when.kind` never active). Each node
//  payload type (`StackProps`, `TextProps`, …) gets its own concrete
//  `applyOverrides` overload; `BuilderNode`'s overload dispatches to the
//  right one and re-wraps the result in the same enum case — mirroring
//  nodes.tsx's `renderNode` calling `applyOverrides(node, ...)` before any
//  per-type rendering.
//
//  Test cases in PaywallOverridesTests.swift mirror validate.test.ts's
//  `applyOverrides` describe block (D-Task 1's shared case table) verbatim.

import Foundation

/// The `{ introEligible, selected }` condition set active for a node's
/// position in the tree. Mirrors nodes.tsx's `activeOverrideConditions`
/// return shape.
public struct OverrideActiveConditions: Equatable, Sendable {
    public let introEligible: Bool
    public let selected: Bool

    public init(introEligible: Bool, selected: Bool) {
        self.introEligible = introEligible
        self.selected = selected
    }
}

/// The active props patches for `overrides`, in array order — matches
/// `active.kind` evaluation in validate.ts's `applyOverrides` (an unknown
/// kind is simply never active, no special-casing needed).
private func activePropPatches<Props>(
    _ overrides: [NodeOverride<Props>]?,
    active: OverrideActiveConditions
) -> [Props] {
    guard let overrides else { return [] }
    return overrides.compactMap { override in
        let isActive: Bool
        switch override.when {
        case .introEligible: isActive = active.introEligible
        case .selected: isActive = active.selected
        case .unknown: isActive = false
        }
        return isActive ? override.props : nil
    }
}

public func applyOverrides(_ props: StackProps, active: OverrideActiveConditions) -> StackProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = StackProps(
            id: result.id, axis: result.axis, children: result.children,
            spacing: patch.spacing ?? result.spacing,
            align: patch.align ?? result.align,
            padding: result.padding, size: result.size,
            background: patch.background ?? result.background,
            cornerRadius: patch.cornerRadius ?? result.cornerRadius,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: TextProps, active: OverrideActiveConditions) -> TextProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = TextProps(
            id: result.id, key: patch.key ?? result.key, role: result.role,
            color: patch.color ?? result.color, align: patch.align ?? result.align,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: ImageProps, active: OverrideActiveConditions) -> ImageProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = ImageProps(
            id: result.id, url: result.url, height: result.height,
            cornerRadius: patch.cornerRadius ?? result.cornerRadius,
            alt: result.alt, overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: ButtonProps, active: OverrideActiveConditions) -> ButtonProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = ButtonProps(
            id: result.id, labelKey: patch.labelKey ?? result.labelKey,
            style: patch.style ?? result.style, action: result.action,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

/// Empty whitelist (`OVERRIDABLE_PROP_KEYS.packageList == []`) — no fields
/// to merge; an `overrides` array on this type can only ever carry
/// `props: {}`, so this is always a no-op.
public func applyOverrides(_ props: PackageListProps, active: OverrideActiveConditions) -> PackageListProps {
    props
}

public func applyOverrides(_ props: PurchaseButtonProps, active: OverrideActiveConditions) -> PurchaseButtonProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = PurchaseButtonProps(
            id: result.id, labelKey: patch.labelKey ?? result.labelKey,
            trialLabelKey: patch.trialLabelKey ?? result.trialLabelKey,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

/// Empty whitelist (`OVERRIDABLE_PROP_KEYS.spacer == []`) — same as
/// `PackageListProps`, always a no-op.
public func applyOverrides(_ props: SpacerProps, active: OverrideActiveConditions) -> SpacerProps {
    props
}

public func applyOverrides(_ props: DividerProps, active: OverrideActiveConditions) -> DividerProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = DividerProps(
            id: result.id, color: patch.color ?? result.color,
            thickness: patch.thickness ?? result.thickness, inset: result.inset,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: IconProps, active: OverrideActiveConditions) -> IconProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = IconProps(
            id: result.id, name: patch.name ?? result.name, size: result.size,
            color: patch.color ?? result.color,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: FeatureListProps, active: OverrideActiveConditions) -> FeatureListProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = FeatureListProps(
            id: result.id, rows: result.rows,
            iconColor: patch.iconColor ?? result.iconColor,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: TimelineProps, active: OverrideActiveConditions) -> TimelineProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = TimelineProps(
            id: result.id, rows: result.rows,
            connectorColor: patch.connectorColor ?? result.connectorColor,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: SocialProofProps, active: OverrideActiveConditions) -> SocialProofProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = SocialProofProps(
            id: result.id, rating: patch.rating ?? result.rating, labelKey: result.labelKey,
            starColor: patch.starColor ?? result.starColor,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: StickyFooterProps, active: OverrideActiveConditions) -> StickyFooterProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = StickyFooterProps(
            id: result.id, children: result.children,
            background: patch.background ?? result.background,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: CountdownProps, active: OverrideActiveConditions) -> CountdownProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = CountdownProps(
            id: result.id, endsAt: result.endsAt, durationSeconds: result.durationSeconds,
            onExpiry: result.onExpiry, labelKey: result.labelKey,
            color: patch.color ?? result.color,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

public func applyOverrides(_ props: CarouselProps, active: OverrideActiveConditions) -> CarouselProps {
    let patches = activePropPatches(props.overrides, active: active)
    guard !patches.isEmpty else { return props }
    var result = props
    for patch in patches {
        result = CarouselProps(
            id: result.id, children: result.children, showsIndicator: result.showsIndicator,
            autoAdvanceSeconds: result.autoAdvanceSeconds, loop: result.loop,
            indicatorColor: patch.indicatorColor ?? result.indicatorColor,
            overrides: result.overrides, fallback: result.fallback)
    }
    return result
}

/// Dispatches to the node's own `applyOverrides` overload and re-wraps the
/// result in the same `BuilderNode` case. `.unknown` nodes carry no
/// overrides field at all and pass through unchanged.
public func applyOverrides(_ node: BuilderNode, active: OverrideActiveConditions) -> BuilderNode {
    switch node {
    case .stack(let p): return .stack(applyOverrides(p, active: active))
    case .text(let p): return .text(applyOverrides(p, active: active))
    case .image(let p): return .image(applyOverrides(p, active: active))
    case .button(let p): return .button(applyOverrides(p, active: active))
    case .packageList(let p): return .packageList(applyOverrides(p, active: active))
    case .purchaseButton(let p): return .purchaseButton(applyOverrides(p, active: active))
    case .spacer(let p): return .spacer(applyOverrides(p, active: active))
    case .divider(let p): return .divider(applyOverrides(p, active: active))
    case .icon(let p): return .icon(applyOverrides(p, active: active))
    case .featureList(let p): return .featureList(applyOverrides(p, active: active))
    case .timeline(let p): return .timeline(applyOverrides(p, active: active))
    case .socialProof(let p): return .socialProof(applyOverrides(p, active: active))
    case .stickyFooter(let p): return .stickyFooter(applyOverrides(p, active: active))
    case .countdown(let p): return .countdown(applyOverrides(p, active: active))
    case .carousel(let p): return .carousel(applyOverrides(p, active: active))
    case .unknown: return node
    }
}

/// The `{ introEligible, selected }` condition set active for a node's
/// position in the tree, given the cell it's scoped to (if any — `nil`
/// outside any `cellTemplate` subtree). Relevance follows the same rule as
/// `{{variable}}` resolution: cell-scoped inside a `cellTemplate` subtree
/// (the cell's own package), selected-scoped everywhere else (the globally
/// selected package). `selected` is only ever true inside a `cellTemplate`
/// subtree, for the cell whose package is the current global selection.
/// Mirrors nodes.tsx's `activeOverrideConditions`.
public func activeOverrideConditions(
    cellPackageId: String?,
    selectedPackageId: String?,
    offering: Offering?
) -> OverrideActiveConditions {
    let relevantPackageId = cellPackageId ?? selectedPackageId
    let introEligible = relevantPackageId.flatMap { id in
        offering?.packages.first(where: { $0.identifier == id })?.product.isEligibleForIntroOffer
    } ?? false
    let selected = cellPackageId != nil && cellPackageId == selectedPackageId
    return OverrideActiveConditions(introEligible: introEligible, selected: selected)
}

File: packages/sdk-kotlin/src/main/kotlin/dev/rovenue/sdk/paywallui/PaywallOverrides.kt
package dev.rovenue.sdk.paywallui

import dev.rovenue.sdk.Offering

// =============================================================
// PaywallOverrides.kt — Kotlin port of
// packages/shared/src/paywall/validate.ts's `applyOverrides`: same
// semantics (array order, later wins, shallow per-field overlay, identity
// when nothing matches, unknown `when.kind` never active). Mirrors
// packages/sdk-swift .../PaywallUI/PaywallOverrides.swift, adapted to the
// Kotlin model where each `BuilderNode` subtype directly carries its own
// fields (no separate `*Props` wrapper the way the Swift enum does).
//
// Test cases in PaywallOverridesTest.kt mirror validate.test.ts's
// `applyOverrides` describe block (D-Task 1's shared case table) verbatim.
// =============================================================

/** The `{ introEligible, selected }` condition set active for a node's
 *  position in the tree. Mirrors nodes.tsx's `activeOverrideConditions`
 *  return shape. */
data class OverrideActiveConditions(val introEligible: Boolean, val selected: Boolean)

/** The active props patches for `overrides`, in array order — matches
 *  `active.kind` evaluation in validate.ts's `applyOverrides` (an unknown
 *  kind is simply never active, no special-casing needed). */
private fun <T> activePropPatches(
    overrides: List<NodeOverride<T>>?,
    active: OverrideActiveConditions,
): List<T> {
    if (overrides == null) return emptyList()
    return overrides.mapNotNull { override ->
        val isActive = when (override.whenKind) {
            OverrideConditionKind.INTRO_ELIGIBLE -> active.introEligible
            OverrideConditionKind.SELECTED -> active.selected
            OverrideConditionKind.UNKNOWN -> false
        }
        if (isActive) override.props else null
    }
}

fun applyOverrides(node: BuilderNode.Stack, active: OverrideActiveConditions): BuilderNode.Stack {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            spacing = patch.spacing ?: result.spacing,
            align = patch.align ?: result.align,
            background = patch.background ?: result.background,
            cornerRadius = patch.cornerRadius ?: result.cornerRadius,
        )
    }
    return result
}

fun applyOverrides(node: BuilderNode.Text, active: OverrideActiveConditions): BuilderNode.Text {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            key = patch.key ?: result.key,
            color = patch.color ?: result.color,
            align = patch.align ?: result.align,
        )
    }
    return result
}

fun applyOverrides(node: BuilderNode.Image, active: OverrideActiveConditions): BuilderNode.Image {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(cornerRadius = patch.cornerRadius ?: result.cornerRadius)
    }
    return result
}

fun applyOverrides(node: BuilderNode.Button, active: OverrideActiveConditions): BuilderNode.Button {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            labelKey = patch.labelKey ?: result.labelKey,
            style = patch.style ?: result.style,
        )
    }
    return result
}

/** Empty whitelist (`OVERRIDABLE_PROP_KEYS.packageList == []`) — no fields
 *  to merge; an `overrides` array on this type can only ever carry
 *  `props: {}`, so this is always a no-op. `active` is unused (kept so the
 *  signature matches every other overload for the generic dispatcher). */
@Suppress("UNUSED_PARAMETER")
fun applyOverrides(node: BuilderNode.PackageList, active: OverrideActiveConditions): BuilderNode.PackageList = node

fun applyOverrides(node: BuilderNode.PurchaseButton, active: OverrideActiveConditions): BuilderNode.PurchaseButton {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            labelKey = patch.labelKey ?: result.labelKey,
            trialLabelKey = patch.trialLabelKey ?: result.trialLabelKey,
        )
    }
    return result
}

/** Empty whitelist (`OVERRIDABLE_PROP_KEYS.spacer == []`) — same as
 *  [BuilderNode.PackageList], always a no-op. */
@Suppress("UNUSED_PARAMETER")
fun applyOverrides(node: BuilderNode.Spacer, active: OverrideActiveConditions): BuilderNode.Spacer = node

fun applyOverrides(node: BuilderNode.Divider, active: OverrideActiveConditions): BuilderNode.Divider {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            color = patch.color ?: result.color,
            thickness = patch.thickness ?: result.thickness,
        )
    }
    return result
}

fun applyOverrides(node: BuilderNode.Icon, active: OverrideActiveConditions): BuilderNode.Icon {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            name = patch.name ?: result.name,
            color = patch.color ?: result.color,
        )
    }
    return result
}

fun applyOverrides(node: BuilderNode.FeatureList, active: OverrideActiveConditions): BuilderNode.FeatureList {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(iconColor = patch.iconColor ?: result.iconColor)
    }
    return result
}

fun applyOverrides(node: BuilderNode.Timeline, active: OverrideActiveConditions): BuilderNode.Timeline {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(connectorColor = patch.connectorColor ?: result.connectorColor)
    }
    return result
}

fun applyOverrides(node: BuilderNode.SocialProof, active: OverrideActiveConditions): BuilderNode.SocialProof {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(
            rating = patch.rating ?: result.rating,
            starColor = patch.starColor ?: result.starColor,
        )
    }
    return result
}

fun applyOverrides(node: BuilderNode.StickyFooter, active: OverrideActiveConditions): BuilderNode.StickyFooter {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(background = patch.background ?: result.background)
    }
    return result
}

fun applyOverrides(node: BuilderNode.Countdown, active: OverrideActiveConditions): BuilderNode.Countdown {
    val patches = activePropPatches(node.overrides, active)
    if (patches.isEmpty()) return node
    var result = node
    for (patch in patches) {
        result = result.copy(color = patch.color ?: result.color)
    }
    return result
}

/** Dispatches to the node's own `applyOverrides` overload. `.Unknown` nodes
 *  carry no overrides field at all and pass through unchanged. */
fun applyOverrides(node: BuilderNode, active: OverrideActiveConditions): BuilderNode = when (node) {
    is BuilderNode.Stack -> applyOverrides(node, active)
    is BuilderNode.Text -> applyOverrides(node, active)
    is BuilderNode.Image -> applyOverrides(node, active)
    is BuilderNode.Button -> applyOverrides(node, active)
    is BuilderNode.PackageList -> applyOverrides(node, active)
    is BuilderNode.PurchaseButton -> applyOverrides(node, active)
    is BuilderNode.Spacer -> applyOverrides(node, active)
    is BuilderNode.Divider -> applyOverrides(node, active)
    is BuilderNode.Icon -> applyOverrides(node, active)
    is BuilderNode.FeatureList -> applyOverrides(node, active)
    is BuilderNode.Timeline -> applyOverrides(node, active)
    is BuilderNode.SocialProof -> applyOverrides(node, active)
    is BuilderNode.StickyFooter -> applyOverrides(node, active)
    is BuilderNode.Countdown -> applyOverrides(node, active)
    is BuilderNode.Unknown -> node
}

/**
 * The `{ introEligible, selected }` condition set active for a node's
 * position in the tree, given the cell it's scoped to (if any — `null`
 * outside any `cellTemplate` subtree). Relevance follows the same rule as
 * `{{variable}}` resolution: cell-scoped inside a `cellTemplate` subtree
 * (the cell's own package), selected-scoped everywhere else (the globally
 * selected package). `selected` is only ever true inside a `cellTemplate`
 * subtree, for the cell whose package is the current global selection.
 * Mirrors nodes.tsx's `activeOverrideConditions`.
 */
fun activeOverrideConditions(
    cellPackageId: String?,
    selectedPackageId: String?,
    offering: Offering?,
): OverrideActiveConditions {
    val relevantPackageId = cellPackageId ?: selectedPackageId
    val introEligible = relevantPackageId?.let { id ->
        offering?.packages?.firstOrNull { it.identifier == id }?.product?.isEligibleForIntroOffer
    } ?: false
    val selected = cellPackageId != null && cellPackageId == selectedPackageId
    return OverrideActiveConditions(introEligible = introEligible, selected = selected)
}
